library/core/src/escape.rs

```diff
@@ -149,77 +151,197 @@ const fn escape_unicode<const N: usize>(c: char) -> ([ascii::Char; N], Range<u8>
+    }
+
     pub(crate) const fn backslash(c: ascii::Char) -> Self {
         let (data, range) = backslash(c);
-        Self { data, alive: range }
+        Self::new(MaybeEscapedCharacter { escaped: data }, range)
     }
 
     pub(crate) const fn ascii(c: u8) -> Self {
         let (data, range) = escape_ascii(c);
-        Self { data, alive: range }
+        Self::new(MaybeEscapedCharacter { escaped: data }, range)
     }
 
     pub(crate) const fn unicode(c: char) -> Self {
         let (data, range) = escape_unicode(c);
-        Self { data, alive: range }
+        Self::new(MaybeEscapedCharacter { escaped: data }, range)
     }
 
     #[inline]
     pub(crate) const fn empty() -> Self {
-        Self { data: [ascii::Char::Null; N], alive: 0..0 }
+        Self::new(MaybeEscapedCharacter { escaped: [ascii::Char::Null; N] }, 0..0)
     }
 
+    /// # Safety
+    ///
+    /// The caller must ensure that `self` contains an escape sequence.
     #[inline]
-    pub(crate) fn as_ascii(&self) -> &[ascii::Char] {
-        // SAFETY: `self.alive` is guaranteed to be a valid range for indexing `self.data`.
+    unsafe fn as_ascii(&self) -> &[ascii::Char] {
+        // SAFETY: `self.data.escaped` contains an escape sequence, as is guaranteed
+        // by the caller, and `self.alive` is guaranteed to be a valid range for
+        // indexing `self.data`.
         unsafe {
-            self.data.get_unchecked(usize::from(self.alive.start)..usize::from(self.alive.end))
+            self.data
+                .escaped
+                .get_unchecked(usize::from(self.alive.start)..usize::from(self.alive.end))
         }
     }
 
+    /// # Safety
+    ///
+    /// The caller must ensure that `self` contains an unescaped `char`.
     #[inline]
-    pub(crate) fn as_str(&self) -> &str {
-        self.as_ascii().as_str()
+    unsafe fn as_char(&self) -> char {
+        // SAFETY: `self.data.unescaped` contains an unescaped `char`,
+        // as is guaranteed by the caller.
+        unsafe { self.data.unescaped }
     }
 
     #[inline]
     pub(crate) fn len(&self) -> usize {
         usize::from(self.alive.end - self.alive.start)
     }
 
+    pub(crate) fn advance_by(&mut self, n: usize) -> Result<(), NonZero<usize>> {
+        self.alive.advance_by(n)
+    }
+
+    pub(crate) fn advance_back_by(&mut self, n: usize) -> Result<(), NonZero<usize>> {
+        self.alive.advance_back_by(n)
+    }
+}
+
+impl<const N: usize> EscapeIterInner<N, AlwaysEscaped> {
     pub(crate) fn next(&mut self) -> Option<u8> {
         let i = self.alive.next()?;
 
-        // SAFETY: `i` is guaranteed to be a valid index for `self.data`.
-        unsafe { Some(self.data.get_unchecked(usize::from(i)).to_u8()) }
+        // SAFETY: We just checked that `self` contains an escape sequence, and
+        // `i` is guaranteed to be a valid index for `self.data.escaped`.
+        unsafe { Some(self.data.escaped.get_unchecked(usize::from(i)).to_u8()) }
```

Comment:
I don't understand the safety comment as written (and checked whether "let i = self.alive.next()?;" does any such checking, but no, it is just a Range<u8> of hopefully valid indices for the escape seq.). Shouldn't this refer to the AlwaysEscaped  type state instead?

Reply:
Yes, it should refer to `AlwaysEscaped` instead.